### PR TITLE
expression: implement vectorized evaluation for `builtinJSONSearchSig`

### PR DESCRIPTION
**Pull request by zhaox1n**
UCP #12104
### What problem does this PR solve?

implement vectorized evaluation for builtinJSOSearchSig

### What is changed and how it works?

```
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-VecBuiltinFunc-12                           1408           1227651 ns/op         1138688 B/op      10240 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-NonVecBuiltinFunc-12                        1076           1158112 ns/op         1138689 B/op      10240 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-VecBuiltinFunc#01-12                         250           4691324 ns/op         2035133 B/op      34837 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-NonVecBuiltinFunc#01-12                      249           4570834 ns/op         2035084 B/op      34837 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-VecBuiltinFunc#02-12                         232           4946785 ns/op         2035053 B/op      34837 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-NonVecBuiltinFunc#02-12                      288           5029420 ns/op         2035164 B/op      34838 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-VecBuiltinFunc#03-12                         278           4720992 ns/op         2067925 B/op      35863 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-NonVecBuiltinFunc#03-12                      241           4428660 ns/op         2068002 B/op      35862 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-VecBuiltinFunc#04-12                         254           4857389 ns/op         2035008 B/op      34837 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-NonVecBuiltinFunc#04-12                      247           4749373 ns/op         2035179 B/op      34838 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-VecBuiltinFunc#05-12                         306           3730100 ns/op         1862828 B/op      21505 allocs/op
BenchmarkVectorizedBuiltinJSONFunc/builtinJSONSearchSig-NonVecBuiltinFunc#05-12                      302           3708237 ns/op         1862739 B/op      21504 allocs/op

```

### Check List 

Tests 

- Unit test

### CODE REVIEW

File: expression/builtin_json_vec.go [hunk 1]
```diff
@@ -394,7 +519,7 @@ func (b *builtinJSONObjectSig) vecEvalJSON(input *chunk.Chunk, result *chunk.Col
 	for i := 0; i < len(b.args); i++ {
 		if i&1 == 0 {
 			if argBuffers[i], err = b.bufAllocator.get(types.ETString, nr); err != nil {
-				return err
+				return nil
```
Comment (wshwsh12) [inline expression/builtin_json_vec.go]: Why return `nil`?

File: expression/builtin_json_vec.go [hunk 2]
```diff
@@ -357,11 +357,136 @@ func (b *builtinJSONQuoteSig) vecEvalString(input *chunk.Chunk, result *chunk.Co
...
+				result.AppendJSON(bj)
+			}
+		}
+	} else {
+		for i := 0; i < nr; i++ {
+			if jsonBuf.IsNull(i) {
+				result.AppendNull()
+				continue
+			}
+			containType := strings.ToLower(typeBuf.GetString(i))
```
Comment (wshwsh12) [inline expression/builtin_json_vec.go]: I see the non-vec function didn't call the strings.ToLower()? It is necessary? If yes, please fix the non-vec function conveniently.

Comment (wshwsh12) [inline expression/builtin_json_vec.go]: Should we check if the `containType` is null?

File: expression/builtin_json_vec.go [hunk 3]
```diff
@@ -357,11 +357,136 @@ func (b *builtinJSONQuoteSig) vecEvalString(input *chunk.Chunk, result *chunk.Co
...
+			escape := byte('\\')
+			bj, err := jsonBuf.GetJSON(i).Search(containType, searchBuf.GetString(i), escape, nil)
```
Comment (wshwsh12) [inline expression/builtin_json_vec.go]: `searchBuf` should be checked, too.

Comment (zhaox1n) [reply expression/builtin_json_vec.go]: All that mentioned has fixed  PTAL @wshwsh12

**Comment by zhaox1n** [comment]
> I think we should use the following code struct to make code clean. Or any idea else? @qw4990
> 
> ```
> for (i := 0;i < nr;i++){
>     if len(b.args) >= 3 {
>         .......
>         if len(b.args) >= 4 {
>             ........
>             if len(b.args) >=5 {
>             ........
>             }
>         }
>     }
> }
> ```
If implementation as you mention it may be 
```
// Initialize variables
 if len(b.args) >= 3 {
        .......
        if len(b.args) >= 4 {
            ........
            if len(b.args) >=5 {
            ........
            }
        }
  }

for (i := 0;i < nr;i++){
    if len(b.args) >= 3 {
        .......
        if len(b.args) >= 4 {
            ........
            if len(b.args) >=5 {
            ........
            }
        }
    }
}
```
Is it ok to create extra `args3` ,`args4`,`args5` function for json_search ?

File: expression/builtin_json_vec.go [hunk 4]
```diff
@@ -357,11 +357,141 @@ func (b *builtinJSONQuoteSig) vecEvalString(input *chunk.Chunk, result *chunk.Co
...
+				} else {
+					return errIncorrectArgs.GenWithStackByArgs("ESCAPE")
+				}
+			}
+			bj, _, err := jsonBuf.GetJSON(i).Search(containType, searchBuf.GetString(i), escape, nil)
+			if err != nil {
+				return err
+			}
+			result.AppendJSON(bj)
+		}
```
Comment (qw4990) [inline expression/builtin_json_vec.go:478]: Sorry for my late reply @zhaox1n .
How about merging these two if-branches together, which would reduce some lines of code?

File: expression/builtin_json_vec.go [hunk 5]
```diff
@@ -357,11 +357,141 @@ func (b *builtinJSONQuoteSig) vecEvalString(input *chunk.Chunk, result *chunk.Co
...
+	} else if len(b.args) == 4 {
+		escapeBuf, err := b.bufAllocator.get(types.ETString, nr)
+		if err != nil {
+			return err
+		}
+		if err := b.args[3].VecEvalString(b.ctx, input, escapeBuf); err != nil {
+			return nil
+		}
+		defer b.bufAllocator.put(escapeBuf)
```
Comment (qw4990) [inline expression/builtin_json_vec.go:455]: Should be:
```
		defer b.bufAllocator.put(escapeBuf)
		if err := b.args[3].VecEvalString(b.ctx, input, escapeBuf); err != nil {
			return nil
		}
```

**Comment by qw4990** [comment]
> > I think we should use the following code struct to make code clean. Or any idea else? @qw4990
> > ```
> > for (i := 0;i < nr;i++){
> >     if len(b.args) >= 3 {
> >         .......
> >         if len(b.args) >= 4 {
> >             ........
> >             if len(b.args) >=5 {
> >             ........
> >             }
> >         }
> >     }
> > }
> > ```
> 
> If implementation as you mention it may be
> 
> ```
> // Initialize variables
>  if len(b.args) >= 3 {
>         .......
>         if len(b.args) >= 4 {
>             ........
>             if len(b.args) >=5 {
>             ........
>             }
>         }
>   }
> 
> for (i := 0;i < nr;i++){
>     if len(b.args) >= 3 {
>         .......
>         if len(b.args) >= 4 {
>             ........
>             if len(b.args) >=5 {
>             ........
>             }
>         }
>     }
> }
> ```
> 
> Is it ok to create extra `args3` ,`args4`,`args5` function for json_search ?

Sorry for my late reply @zhaox1n .
How about:
```
jsonBuf := args[0].VecEvalJSON()
typeBuf := args[1].VecEvalString()
searchBuf := args[2].VecEvalString()

var escapeBuf = nil
if len(args) >= 4 {
	escapeBuf = args[3].VecEvalString()
}

var pathBufs = nil
if len(args) >= 5 {
	pathBufs = append(pathBufs, args[4].VecEvalString(), args[5].VecEvalString() ...)
}

for i := 0; i < nr; i++ {
	if jsonBuf.IsNull(i) || searchBuf.IsNull(i) || typeBuf.IsNull(i) {
		// ...
	}
	for _, path := range pathBufs {
		if path.IsNull(i) {
			// ...
		}
	}
	
	escape = byte('\\')
	if escapeBuf != nil {
		// change escape ...
	}

	// parse path

	// Search
	jsonBuf.GetJSON(i).Search(...)
}
```

**Comment by zhaox1n** [comment]
PTAL @qw4990

**Pull request merged**